File: src/main/java/no/nav/bidrag/beregn/felles/bidragsevne/beregning/BidragsevneberegningImpl.java
package no.nav.bidrag.beregn.felles.bidragsevne.beregning;

import java.util.ArrayList;
import java.util.List;
import no.nav.bidrag.beregn.felles.SjablonUtil;
import no.nav.bidrag.beregn.felles.bidragsevne.bo.BeregnBidragsevneGrunnlagPeriodisert;
import no.nav.bidrag.beregn.felles.bidragsevne.bo.Inntekt;
import no.nav.bidrag.beregn.felles.bidragsevne.bo.ResultatBeregning;
import no.nav.bidrag.beregn.felles.bo.SjablonNokkel;
import no.nav.bidrag.beregn.felles.enums.BostatusKode;
import no.nav.bidrag.beregn.felles.enums.SaerfradragKode;
import no.nav.bidrag.beregn.felles.enums.SjablonInnholdNavn;
import no.nav.bidrag.beregn.felles.enums.SjablonNavn;
import no.nav.bidrag.beregn.felles.enums.SjablonNokkelNavn;
import no.nav.bidrag.beregn.felles.enums.SjablonTallNavn;

//import com.google.common.base.Preconditions;

public class BidragsevneberegningImpl implements Bidragsevneberegning {

  private List<SjablonNokkel> sjablonNokkelListe = new ArrayList<>();

  @Override
  public ResultatBeregning beregn(
      BeregnBidragsevneGrunnlagPeriodisert beregnBidragsevneGrunnlagPeriodisert) {

    //  Preconditions.checkNotNull(beregnBidragsevneGrunnlagPeriodisert, "Grunnlag kan ikke være null");

    System.out.println("Start beregning av bidragsevne");

    Double minstefradrag = beregnMinstefradrag(beregnBidragsevneGrunnlagPeriodisert);

    // Legger sammen inntektene
    var inntekt = beregnBidragsevneGrunnlagPeriodisert.getInntektListe().stream()
        .map(Inntekt::getInntektBelop)
        .reduce(Double.valueOf(0), Double::sum);

    System.out.println("Samlede inntekter: " + inntekt);

    // finner personfradragklasse ut fra angitt skatteklasse
    Double personfradrag = 0.0;
    if (beregnBidragsevneGrunnlagPeriodisert.getSkatteklasse() == (1)) {
      personfradrag = SjablonUtil.hentSjablontall(beregnBidragsevneGrunnlagPeriodisert.getSjablonListe(),
          SjablonTallNavn.PERSONFRADRAG_KLASSE1_BELOP.getNavn(), SjablonInnholdNavn.SJABLON_VERDI.getNavn());
    } else {
      personfradrag = SjablonUtil.hentSjablontall(beregnBidragsevneGrunnlagPeriodisert.getSjablonListe(),
          SjablonTallNavn.PERSONFRADRAG_KLASSE2_BELOP.getNavn(), SjablonInnholdNavn.SJABLON_VERDI.getNavn());
    }

    System.out.println("Beregnet personfradrag: " + personfradrag);

    Double inntektMinusFradrag =
        inntekt - minstefradrag - personfradrag;

    // Trekker fra skatt
    Double forelopigBidragsevne = inntekt - (inntektMinusFradrag
        * SjablonUtil.hentSjablontall(beregnBidragsevneGrunnlagPeriodisert.getSjablonListe(),
        SjablonTallNavn.SKATTESATS_ALMINNELIG_INNTEKT_PROSENT.getNavn(),
        SjablonInnholdNavn.SJABLON_VERDI.getNavn()) / 100);

    System.out.println("Foreløpig evne etter fratrekk av ordinær skatt (totalt + månedlig beløp) : "
        + forelopigBidragsevne + " " + (Double.valueOf(Math.round(forelopigBidragsevne / 12))));

    // Trekker fra trygdeavgift
    System.out.println("Trygdeavgift: " + (Double.valueOf(Math.round(inntekt
        * (SjablonUtil.hentSjablontall(beregnBidragsevneGrunnlagPeriodisert.getSjablonListe(),
        SjablonTallNavn.TRYGDEAVGIFT_PROSENT.getNavn(), SjablonInnholdNavn.SJABLON_VERDI.getNavn()) / 100)))));

    forelopigBidragsevne =
        (forelopigBidragsevne - (inntekt * (
            SjablonUtil.hentSjablontall(beregnBidragsevneGrunnlagPeriodisert.getSjablonListe(),
                SjablonTallNavn.TRYGDEAVGIFT_PROSENT.getNavn(), SjablonInnholdNavn.SJABLON_VERDI.getNavn()) / 100)));
    System.out.println("Foreløpig evne etter fratrekk av trygdeavgift: " + forelopigBidragsevne);

    // Trekker fra trinnvis skatt
    forelopigBidragsevne -= beregnSkattetrinnBelop(beregnBidragsevneGrunnlagPeriodisert);
    System.out.println("Foreløpig evne etter fratrekk av trinnskatt: " + forelopigBidragsevne);


    // Trekker fra boutgifter og midler til eget underhold
    if (beregnBidragsevneGrunnlagPeriodisert.getBostatusKode().equals(BostatusKode.ALENE)) {
      sjablonNokkelListe.clear();
      sjablonNokkelListe.add(new SjablonNokkel(SjablonNokkelNavn.BOSTATUS.getNavn(), "EN"));

      forelopigBidragsevne -= (
          SjablonUtil.hentSjablonverdi(beregnBidragsevneGrunnlagPeriodisert.getSjablonListe(),
              SjablonNavn.BIDRAGSEVNE.getNavn(),
              sjablonNokkelListe,
              SjablonInnholdNavn.BOUTGIFT_BELOP.getNavn()) * 12);

      System.out.println(
          "Foreløpig evne etter fratrekk av boutgifter bor alene: " + forelopigBidragsevne);

      forelopigBidragsevne -= (
          SjablonUtil.hentSjablonverdi(beregnBidragsevneGrunnlagPeriodisert.getSjablonListe(),
              SjablonNavn.BIDRAGSEVNE.getNavn(),
              sjablonNokkelListe,
              SjablonInnholdNavn.UNDERHOLD_BELOP.getNavn()) * 12);

      System.out.println(
          "Foreløpig evne etter fratrekk av midler til eget underhold bor alene: "
              + forelopigBidragsevne);

    } else {
      sjablonNokkelListe.clear();
      sjablonNokkelListe.add(new SjablonNokkel(SjablonNokkelNavn.BOSTATUS.getNavn(), "GS"));

      forelopigBidragsevne -= (
          SjablonUtil.hentSjablonverdi(beregnBidragsevneGrunnlagPeriodisert.getSjablonListe(),
              SjablonNavn.BIDRAGSEVNE.getNavn(),
              sjablonNokkelListe,
              SjablonInnholdNavn.BOUTGIFT_BELOP.getNavn()) * 12);

      System.out.println(
          "Foreløpig evne etter fratrekk av boutgifter gift/samboer: " + forelopigBidragsevne);

      forelopigBidragsevne -= (
          SjablonUtil.hentSjablonverdi(beregnBidragsevneGrunnlagPeriodisert.getSjablonListe(),
              SjablonNavn.BIDRAGSEVNE.getNavn(),
              sjablonNokkelListe,
              SjablonInnholdNavn.UNDERHOLD_BELOP.getNavn()) * 12);

      System.out.println(
          "Foreløpig evne etter fratrekk av midler til eget underhold gift/samboer: "
              + forelopigBidragsevne);
    }

    // Trekker fra midler til underhold egne barn i egen husstand
    forelopigBidragsevne -= (SjablonUtil.hentSjablontall(
        beregnBidragsevneGrunnlagPeriodisert.getSjablonListe(),
        SjablonTallNavn.UNDERHOLD_EGNE_BARN_I_HUSSTAND_BELOP.getNavn(),
        SjablonInnholdNavn.SJABLON_VERDI.getNavn())
            * beregnBidragsevneGrunnlagPeriodisert.getAntallEgneBarnIHusstand()
            * 12);
    System.out.println("Foreløpig evne etter fratrekk av underhold for egne barn i egen husstand: "
        + forelopigBidragsevne);

    // Sjekker om og kalkulerer eventuell fordel særfradrag
    if (beregnBidragsevneGrunnlagPeriodisert.getSaerfradragkode().equals(SaerfradragKode.HELT)) {
      forelopigBidragsevne += SjablonUtil.hentSjablontall(
          beregnBidragsevneGrunnlagPeriodisert.getSjablonListe(),
          SjablonTallNavn.FORDEL_SAERFRADRAG_BELOP.getNavn(),
          SjablonInnholdNavn.SJABLON_VERDI.getNavn());

      System.out.println("Foreløpig evne etter tillegg for særfradrag: " + forelopigBidragsevne);
    } else {
      if (beregnBidragsevneGrunnlagPeriodisert.getSaerfradragkode().equals(SaerfradragKode.HALVT)) {
        forelopigBidragsevne += (
            SjablonUtil.hentSjablontall(
                beregnBidragsevneGrunnlagPeriodisert.getSjablonListe(),
                SjablonTallNavn.FORDEL_SAERFRADRAG_BELOP.getNavn(),
                SjablonInnholdNavn.SJABLON_VERDI.getNavn()) / 2);
        System.out
            .println("Foreløpig evne etter tillegg for halvt særfradrag: " + forelopigBidragsevne);
      }
    }

    // Legger til fordel skatteklasse2
    forelopigBidragsevne += SjablonUtil.hentSjablontall(
        beregnBidragsevneGrunnlagPeriodisert.getSjablonListe(),
        SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELOP.getNavn(),
        SjablonInnholdNavn.SJABLON_VERDI.getNavn());
    System.out
        .println("Foreløpig evne etter tillegg for fordel skatteklasse2: " + forelopigBidragsevne);

    // Finner månedlig beløp for bidragsevne
    Double maanedligBidragsevne = Double.valueOf(Math.round(forelopigBidragsevne / 12));
    System.out.println("Endelig beregnet bidragsevne: " + maanedligBidragsevne);

    if (maanedligBidragsevne.compareTo(0.0) < 0){
      System.out.println("Beregnet bidragsevne er mindre enn 0, settes til 0");
      maanedligBidragsevne = Double.valueOf(0.0);
      System.out.println("Korrigert bidragsevne: " + maanedligBidragsevne);
    }
    System.out.println("------------------------------------------------------");

    return new ResultatBeregning(maanedligBidragsevne);

  }

  @Override
  public Double beregnMinstefradrag(
      BeregnBidragsevneGrunnlagPeriodisert beregnBidragsevneGrunnlagPeriodisert) {

    // Legger sammen inntektene
    var inntekt = beregnBidragsevneGrunnlagPeriodisert.getInntektListe().stream()
        .map(Inntekt::getInntektBelop)
        .reduce(Double.valueOf(0), Double::sum);

    Double minstefradrag = inntekt * (
        SjablonUtil.hentSjablontall(
            beregnBidragsevneGrunnlagPeriodisert.getSjablonListe(),
            SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT.getNavn(),
            SjablonInnholdNavn.SJABLON_VERDI.getNavn()) / 100);
    if (minstefradrag.compareTo(
        SjablonUtil.hentSjablontall(
            beregnBidragsevneGrunnlagPeriodisert.getSjablonListe(),
            SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELOP.getNavn(),
            SjablonInnholdNavn.SJABLON_VERDI.getNavn())) > 0) {
      minstefradrag = SjablonUtil.hentSjablontall(
          beregnBidragsevneGrunnlagPeriodisert.getSjablonListe(),
          SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELOP.getNavn(),
          SjablonInnholdNavn.SJABLON_VERDI.getNavn());
    }
    System.out.println("Beregnet minstefradrag: " + minstefradrag);
    return minstefradrag;
  }

  @Override
  public Double beregnSkattetrinnBelop(
      BeregnBidragsevneGrunnlagPeriodisert beregnBidragsevneGrunnlagPeriodisert) {

    // Legger sammen inntektene
    var inntekt = beregnBidragsevneGrunnlagPeriodisert.getInntektListe().stream()
        .map(Inntekt::getInntektBelop)
        .reduce(Double.valueOf(0), Double::sum);

    var sortertTrinnvisSkattesatsListe = SjablonUtil.hentTrinnvisSkattesats(
        beregnBidragsevneGrunnlagPeriodisert.getSjablonListe(),
        SjablonNavn.TRINNVIS_SKATTESATS.getNavn());

    var samletSkattetrinnBelop = 0d;
    var indeks = 1;

    // Beregn skattetrinnbeløp
    while (indeks < sortertTrinnvisSkattesatsListe.size()) {
      if (inntekt > sortertTrinnvisSkattesatsListe.get(indeks - 1).getInntektGrense()) {
        if (inntekt < sortertTrinnvisSkattesatsListe.get(indeks).getInntektGrense()) {
          samletSkattetrinnBelop = Math.round(samletSkattetrinnBelop + (
              (inntekt - sortertTrinnvisSkattesatsListe.get(indeks - 1).getInntektGrense()) *
                  (sortertTrinnvisSkattesatsListe.get(indeks - 1).getSats() / 100)));
        } else {
          samletSkattetrinnBelop = Math.round(samletSkattetrinnBelop + (
              (sortertTrinnvisSkattesatsListe.get(indeks).getInntektGrense() -
                  sortertTrinnvisSkattesatsListe.get(indeks - 1).getInntektGrense()) * (
                  sortertTrinnvisSkattesatsListe.get(indeks - 1).getSats() / 100)));
        }
      }
      indeks = indeks + 1;
    }

    if (inntekt > sortertTrinnvisSkattesatsListe.get(indeks - 1).getInntektGrense()) {
      samletSkattetrinnBelop = Math.round(samletSkattetrinnBelop + (
          (inntekt - sortertTrinnvisSkattesatsListe.get(indeks - 1).getInntektGrense())
              * (sortertTrinnvisSkattesatsListe.get(indeks - 1).getSats() / 100)));
    }


    System.out.println("Totalt skattetrinnbeløp: " + samletSkattetrinnBelop);

    return samletSkattetrinnBelop;
  }
}

File: bidrag-beregn-særtilskudd/src/main/kotlin/no/nav/bidrag/beregn/bidragsevne/beregning/BidragsevneBeregningImpl.kt
package no.nav.bidrag.beregn.bidragsevne.beregning

import no.nav.bidrag.beregn.bidragsevne.bo.GrunnlagBeregning
import no.nav.bidrag.beregn.bidragsevne.bo.Inntekt
import no.nav.bidrag.beregn.bidragsevne.bo.ResultatBeregning
import no.nav.bidrag.beregn.felles.FellesBeregning
import no.nav.bidrag.beregn.felles.bo.SjablonNokkel
import no.nav.bidrag.beregn.felles.bo.SjablonPeriode
import no.nav.bidrag.beregn.felles.util.SjablonUtil
import no.nav.bidrag.domene.enums.beregning.Særfradragskode
import no.nav.bidrag.domene.enums.person.Bostatuskode
import no.nav.bidrag.domene.enums.sjablon.SjablonInnholdNavn
import no.nav.bidrag.domene.enums.sjablon.SjablonNavn
import no.nav.bidrag.domene.enums.sjablon.SjablonNøkkelNavn
import no.nav.bidrag.domene.enums.sjablon.SjablonTallNavn
import java.math.BigDecimal
import java.math.MathContext
import java.math.RoundingMode

class BidragsevneBeregningImpl : FellesBeregning(), BidragsevneBeregning {

    override fun beregn(grunnlag: GrunnlagBeregning): ResultatBeregning {
        // Henter sjablonverdier
        val sjablonNavnVerdiMap = hentSjablonVerdier(
            sjablonPeriodeListe = grunnlag.sjablonListe,
            bostatusKode = grunnlag.bostatus.kode,
            skatteklasse = grunnlag.skatteklasse.skatteklasse,
        )

        // Beregner minstefradrag
        val minstefradrag = beregnMinstefradrag(
            grunnlag = grunnlag,
            minstefradragInntektSjablonBelop = (sjablonNavnVerdiMap[SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELØP.navn])!!,
            minstefradragInntektSjablonProsent = (sjablonNavnVerdiMap[SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT.navn])!!,
        )

        // Legger sammen inntektene
        val inntekt = grunnlag.inntektListe.stream()
            .map(Inntekt::inntektBelop)
            .reduce(BigDecimal.ZERO, BigDecimal::add)

        // Finner personfradragklasse ut fra angitt skatteklasse
        val personfradrag = if (grunnlag.skatteklasse.skatteklasse == 1) {
            sjablonNavnVerdiMap[SjablonTallNavn.PERSONFRADRAG_KLASSE1_BELØP.navn]
        } else {
            sjablonNavnVerdiMap[SjablonTallNavn.PERSONFRADRAG_KLASSE2_BELØP.navn]
        }

        val inntektMinusFradrag = inntekt.subtract(minstefradrag).subtract(personfradrag)

        var forelopigBidragsevne = inntekt
            // Trekker fra skatt
            .subtract(
                inntektMinusFradrag.multiply(
                    sjablonNavnVerdiMap[SjablonTallNavn.SKATTESATS_ALMINNELIG_INNTEKT_PROSENT.navn]!!.divide(
                        BigDecimal.valueOf(100),
                        MathContext(10, RoundingMode.HALF_UP),
                    ),
                ),
            )
            .subtract(
                inntekt.multiply(
                    sjablonNavnVerdiMap[SjablonTallNavn.TRYGDEAVGIFT_PROSENT.navn]!!.divide(
                        BigDecimal.valueOf(100),
                        MathContext(10, RoundingMode.HALF_UP),
                    ),
                ),
            )
            // Trekker fra trinnvis skatt
            .subtract(beregnSkattetrinnBelop(grunnlag))
            // Trekker fra boutgifter og midler til eget underhold
            .subtract(sjablonNavnVerdiMap[SjablonInnholdNavn.BOUTGIFT_BELØP.navn]!!.multiply(BigDecimal.valueOf(12)))
            .subtract(sjablonNavnVerdiMap[SjablonInnholdNavn.UNDERHOLD_BELØP.navn]!!.multiply(BigDecimal.valueOf(12)))
            // Trekker fra midler til underhold egne barn i egen husstand
            .subtract(
                sjablonNavnVerdiMap[SjablonTallNavn.UNDERHOLD_EGNE_BARN_I_HUSSTAND_BELØP.navn]!!
                    .multiply(BigDecimal.valueOf(grunnlag.barnIHusstand.antallBarn))
                    .multiply(BigDecimal.valueOf(12)),
            )

        // Sjekker om og kalkulerer eventuell fordel særfradrag
        val fordelSaerfradragBelop = sjablonNavnVerdiMap[SjablonTallNavn.FORDEL_SÆRFRADRAG_BELØP.navn]
        forelopigBidragsevne = when (grunnlag.saerfradrag.kode) {
            Særfradragskode.HELT -> forelopigBidragsevne + (fordelSaerfradragBelop ?: BigDecimal.ZERO)
            Særfradragskode.HALVT -> {
                val halvVerdi = fordelSaerfradragBelop?.divide(BigDecimal.valueOf(2), MathContext(10, RoundingMode.HALF_UP))
                forelopigBidragsevne + (halvVerdi ?: BigDecimal.ZERO)
            }

            else -> forelopigBidragsevne
        }

        // Legger til fordel skatteklasse2
        if (grunnlag.skatteklasse.skatteklasse == 2) {
            forelopigBidragsevne = forelopigBidragsevne.add(sjablonNavnVerdiMap[SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELØP.navn])
        }

        // Finner månedlig beløp for bidragsevne
        val maanedligBidragsevne = maxOf(
            forelopigBidragsevne.divide(BigDecimal.valueOf(12), MathContext(10, RoundingMode.HALF_UP)).setScale(0, RoundingMode.HALF_UP),
            BigDecimal.ZERO,
        )

        return ResultatBeregning(
            belop = maanedligBidragsevne,
            sjablonListe = byggSjablonResultatListe(sjablonNavnVerdiMap = sjablonNavnVerdiMap, sjablonPeriodeListe = grunnlag.sjablonListe),
        )
    }

    override fun beregnMinstefradrag(
        grunnlag: GrunnlagBeregning,
        minstefradragInntektSjablonBelop: BigDecimal,
        minstefradragInntektSjablonProsent: BigDecimal,
    ): BigDecimal {
        // Legger sammen inntektene
        val inntekt = grunnlag.inntektListe.stream()
            .map(Inntekt::inntektBelop)
            .reduce(BigDecimal.ZERO, BigDecimal::add)

        return minOf(
            inntekt.multiply(minstefradragInntektSjablonProsent.divide(BigDecimal.valueOf(100), MathContext(2, RoundingMode.HALF_UP))),
            minstefradragInntektSjablonBelop,
        ).setScale(0, RoundingMode.HALF_UP)
    }

    override fun beregnSkattetrinnBelop(grunnlag: GrunnlagBeregning): BigDecimal {
        // Legger sammen inntektene
        val inntekt = grunnlag.inntektListe.stream()
            .map(Inntekt::inntektBelop)
            .reduce(BigDecimal.ZERO, BigDecimal::add)

        val sortertTrinnvisSkattesatsListe = SjablonUtil.hentTrinnvisSkattesats(
            sjablonListe = grunnlag.sjablonListe.map { it.sjablon }.toList(),
            sjablonNavn = SjablonNavn.TRINNVIS_SKATTESATS,
        )

        var samletSkattetrinnBelop = BigDecimal.ZERO
        var indeks = 1

        // Beregner skattetrinnbeløp
        while (indeks < sortertTrinnvisSkattesatsListe.size) {
            val denneGrense = sortertTrinnvisSkattesatsListe[indeks - 1].inntektGrense
            val nesteGrense = sortertTrinnvisSkattesatsListe[indeks].inntektGrense

            if (inntekt > denneGrense) {
                val taxableIncome = minOf(inntekt, nesteGrense) - denneGrense
                val taxRate = sortertTrinnvisSkattesatsListe[indeks - 1].sats

                samletSkattetrinnBelop += (taxableIncome * taxRate).divide(BigDecimal.valueOf(100), MathContext(10, RoundingMode.HALF_UP))
            }
            indeks++
        }

        if (inntekt > sortertTrinnvisSkattesatsListe[indeks - 1].inntektGrense) {
            val taxableIncome = inntekt - sortertTrinnvisSkattesatsListe[indeks - 1].inntektGrense
            val taxRate = sortertTrinnvisSkattesatsListe[indeks - 1].sats

            samletSkattetrinnBelop += (taxableIncome * taxRate / BigDecimal.valueOf(100)).setScale(1, RoundingMode.HALF_UP)
        }

        return samletSkattetrinnBelop.setScale(0, RoundingMode.HALF_UP)
    }

    // Henter sjablonverdier
    private fun hentSjablonVerdier(
        sjablonPeriodeListe: List<SjablonPeriode>,
        bostatusKode: Bostatuskode,
        skatteklasse: Int,
    ): Map<String, BigDecimal> {
        val sjablonNavnVerdiMap = HashMap<String, BigDecimal>()
        val sjablonListe = sjablonPeriodeListe.map { it.sjablon }.toList()

        // Sjablontall
        if (skatteklasse == 1) {
            sjablonNavnVerdiMap[SjablonTallNavn.PERSONFRADRAG_KLASSE1_BELØP.navn] =
                SjablonUtil.hentSjablonverdi(sjablonListe = sjablonListe, sjablonTallNavn = SjablonTallNavn.PERSONFRADRAG_KLASSE1_BELØP)
        } else {
            sjablonNavnVerdiMap[SjablonTallNavn.PERSONFRADRAG_KLASSE2_BELØP.navn] =
                SjablonUtil.hentSjablonverdi(sjablonListe = sjablonListe, sjablonTallNavn = SjablonTallNavn.PERSONFRADRAG_KLASSE2_BELØP)
            sjablonNavnVerdiMap[SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELØP.navn] =
                SjablonUtil.hentSjablonverdi(sjablonListe = sjablonListe, sjablonTallNavn = SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELØP)
        }
        sjablonNavnVerdiMap[SjablonTallNavn.SKATTESATS_ALMINNELIG_INNTEKT_PROSENT.navn] =
            SjablonUtil.hentSjablonverdi(sjablonListe = sjablonListe, sjablonTallNavn = SjablonTallNavn.SKATTESATS_ALMINNELIG_INNTEKT_PROSENT)
        sjablonNavnVerdiMap[SjablonTallNavn.TRYGDEAVGIFT_PROSENT.navn] =
            SjablonUtil.hentSjablonverdi(sjablonListe = sjablonListe, sjablonTallNavn = SjablonTallNavn.TRYGDEAVGIFT_PROSENT)
        sjablonNavnVerdiMap[SjablonTallNavn.UNDERHOLD_EGNE_BARN_I_HUSSTAND_BELØP.navn] =
            SjablonUtil.hentSjablonverdi(sjablonListe = sjablonListe, sjablonTallNavn = SjablonTallNavn.UNDERHOLD_EGNE_BARN_I_HUSSTAND_BELØP)
        sjablonNavnVerdiMap[SjablonTallNavn.FORDEL_SÆRFRADRAG_BELØP.navn] =
            SjablonUtil.hentSjablonverdi(sjablonListe = sjablonListe, sjablonTallNavn = SjablonTallNavn.FORDEL_SÆRFRADRAG_BELØP)
        sjablonNavnVerdiMap[SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT.navn] =
            SjablonUtil.hentSjablonverdi(sjablonListe = sjablonListe, sjablonTallNavn = SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT)
        sjablonNavnVerdiMap[SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELØP.navn] =
            SjablonUtil.hentSjablonverdi(sjablonListe = sjablonListe, sjablonTallNavn = SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELØP)

        // Bidragsevne
        val sjablonNokkelVerdi = if ((bostatusKode == Bostatuskode.ALENE)) "EN" else "GS"
        sjablonNavnVerdiMap[SjablonInnholdNavn.BOUTGIFT_BELØP.navn] = SjablonUtil.hentSjablonverdi(
            sjablonListe = sjablonListe,
            sjablonNavn = SjablonNavn.BIDRAGSEVNE,
            sjablonNokkelListe = listOf(SjablonNokkel(navn = SjablonNøkkelNavn.BOSTATUS.navn, verdi = sjablonNokkelVerdi)),
            sjablonInnholdNavn = SjablonInnholdNavn.BOUTGIFT_BELØP,
        )
        sjablonNavnVerdiMap[SjablonInnholdNavn.UNDERHOLD_BELØP.navn] = SjablonUtil.hentSjablonverdi(
            sjablonListe = sjablonListe,
            sjablonNavn = SjablonNavn.BIDRAGSEVNE,
            sjablonNokkelListe = listOf(SjablonNokkel(navn = SjablonNøkkelNavn.BOSTATUS.navn, verdi = sjablonNokkelVerdi)),
            sjablonInnholdNavn = SjablonInnholdNavn.UNDERHOLD_BELØP,
        )

        // TrinnvisSkattesats
        val trinnvisSkattesatsListe =
            SjablonUtil.hentTrinnvisSkattesats(sjablonListe = sjablonListe, sjablonNavn = SjablonNavn.TRINNVIS_SKATTESATS)
        var indeks = 1
        trinnvisSkattesatsListe.forEach {
            sjablonNavnVerdiMap[SjablonNavn.TRINNVIS_SKATTESATS.navn + "InntektGrense" + indeks] = it.inntektGrense
            sjablonNavnVerdiMap[SjablonNavn.TRINNVIS_SKATTESATS.navn + "Sats" + indeks] = it.sats
            indeks++
        }

        return sjablonNavnVerdiMap
    }
}
